prefect_azure/container_instance.py

```diff
@@ -0,0 +1,704 @@
+            The time of the most recent output line written by this call.
+        """
+        if not log_content:
+            # nothing to stream
+            return last_log_time
+
+        log_lines = log_content.split("\n")
+
+        last_written_time = last_log_time
+
+        for log_line in log_lines:
+            # skip if the line is blank or whitespace
+            if not log_line.strip():
+                continue
+
+            line_parts = log_line.split(" ")
+            # timestamp should always be before first space in line
+            line_timestamp = line_parts[0]
+            line = " ".join(line_parts[1:])
+
+            try:
+                line_time = dateutil.parser.parse(line_timestamp)
+                if line_time > last_written_time:
+                    self._write_output_line(line)
+                    last_written_time = line_time
+            except dateutil.parser.ParserError as e:
+                self.logger.debug(
+                    "Unable to parse timestamp from Azure log line.",
+                    log_line,
+                    exc_info=e,
+                )
+
+        return last_written_time
+
+    def _create_credential(self):
+        """
+        Creates an Azure credential initialized with data from this block's fields.
+
+        Returns:
+            An initialized Azure `TokenCredential` ready to use with Azure SDK client
+            classes.
+        """
+        return ClientSecretCredential(
+            tenant_id=self.aci_credentials.tenant_id,
+            client_id=self.aci_credentials.client_id,
+            client_secret=self.aci_credentials.client_secret.get_secret_value(),
+        )
+
+    def _create_container_client(self, token_credential: TokenCredential):
+        """
+        Creates an Azure Container Instances client initialized with data from
+        this block's fields.
+
+        Args:
+            token_credential: A valid Azure `TokenCredential`
+
+        Returns:
+            An initialized `ContainerInstanceManagementClient`
+        """
+
+        return ContainerInstanceManagementClient(
+            credential=token_credential,
+            subscription_id=self.subscription_id.get_secret_value(),
+        )
+
+    def _create_resource_client(self, token_credential: TokenCredential):
+        """
+        Creates an Azure resource management client initialized with data from
+        this block's fields.
+
+        Args:
+            token_credential: A valid Azure `TokenCredential`
+
+        Returns:
+            An initialized `ResourceManagementClient`
+        """
+
+        return ResourceManagementClient(
+            credential=token_credential,
+            subscription_id=self.subscription_id.get_secret_value(),
+        )
+
+    @property
+    def _environment(self):
+        """
+        Generates a dictionary of all environment variables to send to the
+        ACI container.
+        """
+        return {**self._base_environment(), **self.env}
```

Comment:
We don't use a property for this in the other infrastructure blocks. I'd use a method for consistency.